include/devicetree.h

```diff
@@ -145,6 +145,15 @@
  */
 #define DT_NODELABEL(label) DT_CAT(DT_N_NODELABEL_, label)
 
+/**
+ * @brief returns list of child node identifiers.
+ * List contains only praimary enabled child nodes (no aliases).
```

Comment:
Since DT_PARENT() returns the parent regardless of if it's enabled, I think this macro should do the same. The user can always skip generating code for a particular child if it's not enabled using something like `COND_CODE_1(DT_HAS_NODE(child), (DO_SOMETHING_WITH(child),), ())`

Reply:
OK

Reply:
Just checking -- this will be part of a later update, right? It doesn't seem done yet when I checked the output. For example, when building hello_world for nRF52840-DK, I see:

```
/* Node children (/) identifiers: */
#define DT_N_CHILD_IDENTIFIERS sw_pwm, arduino_header, arduino_adc
```

But `/` has more children not in that list (like `soc`, `leds`, etc.)

Reply:
Originally I have been going after the labels, the things ending with ':', `soc`, `leds` does not end with `:` . I think these should be on different lists then as someone may use own labels , eg `alarm_led: led_0`

Reply:
Oh, I missed that, I'm sorry. Hrm, this is new, then. We've normally just operated on nodes themselves. Node labels were just a way to get to the nodes.

This starts to feel tricky because a node can have zero, one, or many nodes, so this is potentially looping over child nodes multiple times. I see why the flash partitions work wants to operate on node labels so you can do things like create an enum with 'boot_partition' in the identifier. But I think this solution probably ought to work even with nodes that do not have labels, since their presence is not guaranteed.

Reply:
@mbolivar-nordic  OK, I fix it to generate list of all children by the node identifier, not node label. It will not help me with what I do, but will be useful  anyway, so no reason to waste the effort

Reply:
> It will not help me with what I do, but will be useful anyway, so no reason to waste the effort

How come?

I guess I was figuring something like this (completely untested):

```
#define FLASH_AREA(partition_node_identifier) flash_area_##partition_node_identifier

#define FLASH_MEMORY DT_CHOSEN(zephyr_flash)
#define FLASH_PARTITIONS DT_CHILD(FLASH_MEMORY, partitions)

#define MAKE_FLASH_ENUM(partition_node) FLASH_AREA(partition_node),
enum my_partitions {
    SOME_UTIL_FOR_EACH_MACRO_TBD(MAKE_FLASH_ENUM, DT_CHILDREN(FLASH_PARTITIONS))
};
```

Now, if you want to know the flash enumerator for the boot partition, you can use something like `FLASH_AREA(DT_NODELABEL(boot_partition))`.

What am I missing here?

Reply:
>```
> #define FLASH_AREA(partition_node_identifier) flash_area_##partition_node_identifier
> ...
> FLASH_AREA(DT_NODELABEL(boot_partition))
> ```

Invocations like that do not work, because C preprocessor, when evaluating macro, first does parameter substitution and processes # and ## operators, before starting with replacing macro names with definitions. This means that above will stop at:
```
flash_area_DT_NODELABEL(boot_partition)
```

Reply:
The usual trick to resolve that is to do another layer of expansion, like with 63c936df7d1e6a325717e4554c301bf04c3d4983

Reply:
Thanks, done it.